Add a debug pixel to check the tab preview count

samsymons opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/414235014887631/1200123805767782/f
Tech Design URL:
CC: @bwaresiak 

**Description**:

This PR adds a debug pixel to check the stored tab preview count vs the total tab count, reporting the pixel if the preview count is higher than the tab count.

One question about this change is: are there any situations where we would expect the tab preview count to be higher, even briefly? In the implementation I've added the pixel check after tab previews and the tab model are updated, but I want to be certain that we won't fire false positives.

**Steps to test this PR**:
1. Open and close tabs, checking that the pixel check is performed (but that the pixel does not fire)
1. Check that there are no other places where tab previews are stored, so that we're sure the pixel is checking all cases
1. Background and foreground the app, checking that the pixel check is performed (but that the pixel does not fire)

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable. 
-->

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [ ] iOS 11
* [ ] iOS 12
* [ ] iOS 13
* [ ] iOS 14

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

samsymons committed:
Check the tab preview vs tab count after the model changes.

samsymons committed:
Return an optional Int if the file system can’t get the count.

Better to not report this as a tab preview count of 0, since that could be misleading if the real problem is with the file system.

bwaresiak commented on DuckDuckGo/TabManager.swift:
```diff
@@ -224,6 +224,18 @@ class TabManager {
 
     func save() {
         model.save()
+        assertTabPreviewCount()
```
The only concern I have is how often this is called (especially when loading the page). This is simple, as 'save' is a single point to catch all model changes - which is very good, but on the other hand not all of the changes result in preview creation/deletion causing potentially unnecessary checks..

samsymons committed:
Use a notification to determine when to check the tab count.

samsymons committed:
Remove the tab manager as a notification observer.

samsymons committed:
Undo the deinit change for clarity, it isn’t necessary.

“If your app targets iOS 9.0 and later or macOS 10.11 and later, and you used addObserver(_:selector:name:object:), you do not need to unregister the observer.”

bwaresiak reviewed (approved):
LGTM!
